util/teststorage/appender.go

```diff
@@ -0,0 +1,415 @@
+
+package teststorage
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+)
+
+// Sample represents test, combined sample for mocking storage.AppenderV2.
+type Sample struct {
+	MF    string
+	L     labels.Labels
+	M     metadata.Metadata
+	ST, T int64
+	V     float64
+	H     *histogram.Histogram
+	FH    *histogram.FloatHistogram
+	ES    []exemplar.Exemplar
+}
+
+func (s Sample) String() string {
+	b := bytes.Buffer{}
+	if s.M.Help != "" {
+		_, _ = fmt.Fprintf(&b, "HELP %s\n", s.M.Help)
+	}
+	if s.M.Type != model.MetricTypeUnknown && s.M.Type != "" {
+		_, _ = fmt.Fprintf(&b, "type@%s ", s.M.Type)
+	}
+	if s.M.Unit != "" {
+		_, _ = fmt.Fprintf(&b, "unit@%s ", s.M.Unit)
+	}
+	h := ""
+	if s.H != nil {
+		h = s.H.String()
+	}
+	fh := ""
+	if s.FH != nil {
+		fh = s.FH.String()
+	}
+	_, _ = fmt.Fprintf(&b, "%s %v%v%v st@%v t@%v\n", s.L.String(), s.V, h, fh, s.ST, s.T)
+	return b.String()
+}
+
+func (s Sample) exemplarsEquals(other []exemplar.Exemplar) bool {
+	if len(s.ES) != len(other) {
+		return false
+	}
+	for i := range s.ES {
+		if !s.ES[i].Equals(other[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func (s Sample) Equals(other Sample) bool {
+	return strings.Compare(s.MF, other.MF) == 0 &&
+		labels.Equal(s.L, other.L) &&
+		s.M.Equals(other.M) &&
+		s.ST == other.ST &&
+		s.T == other.T &&
+		math.Float64bits(s.V) == math.Float64bits(other.V) && // Compare Float64bits so NaN values which are exactly the same will compare equal.
+		s.H.Equals(other.H) &&
+		s.FH.Equals(other.FH) &&
+		s.exemplarsEquals(other.ES)
+}
+
+// sample exists to allow using Sample in tests usings chunks.Sample interface.
+// TODO(bwplotka): Consider deduplicate those?
+type sample struct {
+	Sample
+}
+
+var _ chunks.Sample = sample{}
+
+func (s sample) T() int64 { return s.Sample.T }
+
+func (s sample) F() float64 { return s.V }
+
+func (s sample) H() *histogram.Histogram { return s.Sample.H }
+
+func (s sample) FH() *histogram.FloatHistogram { return s.Sample.FH }
+
+func (s sample) Type() chunkenc.ValueType {
+	switch {
+	case s.Sample.H != nil:
+		return chunkenc.ValHistogram
+	case s.Sample.FH != nil:
+		return chunkenc.ValFloatHistogram
+	default:
+		return chunkenc.ValFloat
+	}
+}
+
+func (s sample) Copy() chunks.Sample {
+	c := sample{Sample: Sample{T: s.T(), V: s.F()}}
+	if s.H() != nil {
+		c.Sample.H = s.H().Copy()
+	}
+	if s.FH() != nil {
+		c.Sample.FH = s.FH().Copy()
+	}
+	return c
+}
+
+// ToChunkSampleSlice converts []Samples to []chunks.Sample.
+func ToChunkSampleSlice(samples []Sample) (ret []chunks.Sample) {
```

Comment:
Note to reviewers: We could remove this code as it's only used in `TestConvertClassicHistogramsToNHCB`. If we remove querier select path, we could NOT use SeriesSet. However I think sooner than later someone would need to represent `teststorage.Sample` as `chunks.Sample` and `storage.SeriesSet` etc